crates/cache/src/oci.rs

```diff
@@ -0,0 +1,130 @@
+const OCI_CACHE_DIR: &str = "oci";
+const MANIFESTS_DIR: &str = "manifests";
+const WASM_DIR: &str = "wasm";
+const DATA_DIR: &str = "data";
+
+/// Cache for registry entities.
+pub struct Cache {
+    /// Root directory for the cache instance.
+    pub root: PathBuf,
+}
+
+impl Cache {
+    /// Create a new cache given an optional root directory.
+    pub async fn new(root: Option<PathBuf>) -> Result<Self> {
+        let root = match root {
+            Some(root) => root,
+            None => dirs::config_dir()
+                .context("cannot get configuration directory")?
+                .join(CONFIG_DIR),
+        };
+        let root = root.join(REGISTRY_CACHE_DIR).join(OCI_CACHE_DIR);
+        Self::ensure_dirs(&root).await?;
+
+        Ok(Self { root })
+    }
+
+    /// The manifests directory for the current cache.
+    pub fn manifests_dir(&self) -> PathBuf {
+        self.root.join(MANIFESTS_DIR)
+    }
+
+    /// The Wasm bytes directory for the current cache.
+    pub fn wasm_dir(&self) -> PathBuf {
+        self.root.join(WASM_DIR)
+    }
+
+    /// The data directory for the current cache.
+    pub fn data_dir(&self) -> PathBuf {
+        self.root.join(DATA_DIR)
+    }
+
+    /// Get the file path to a manifest given a reference.
+    /// If the directory for the manifest does not exist, this will create it.
+    pub async fn manifest_for_reference(&self, reference: &str) -> Result<PathBuf> {
+        let reference: Reference = reference.parse().context("cannot parse reference")?;
+        let p = self
+            .manifests_dir()
+            .join(reference.registry())
+            .join(reference.repository())
+            .join(reference.tag().unwrap_or("latest"));
+
+        if !p.is_dir() {
+            fs::create_dir_all(&p).await?;
+        }
+
+        Ok(p.join("manifest.json"))
+    }
+
+    /// Get the file path to a config object given a reference.
+    pub async fn config_for_reference(&self, reference: &str) -> Result<PathBuf> {
+        let reference: Reference = reference.parse().context("cannot parse reference")?;
+        let p = self
+            .manifests_dir()
+            .join(reference.registry())
+            .join(reference.repository())
+            .join(reference.tag().unwrap_or("latest"));
+
+        if !p.is_dir() {
+            fs::create_dir_all(&p).await?;
+        }
+
+        Ok(p.join("config.json"))
+    }
+
+    /// Write the contents in the cache's wasm directory.
+    pub async fn write_wasm(&self, bytes: &Vec<u8>, digest: &str) -> Result<()> {
+        fs::write(self.wasm_dir().join(digest), bytes).await?;
+        Ok(())
+    }
+
+    /// Write the contents in the cache's data directory.
+    pub async fn write_data(&self, bytes: &Vec<u8>, digest: &str) -> Result<()> {
+        fs::write(self.data_dir().join(digest), bytes).await?;
+        Ok(())
+    }
+
+    /// Ensure the expected configuration directories are found in the root.
+    /// └── fermyon
```

Comment:
Is this using `fermyon` to stand for the root, or is `fermyon` under the root?  (Thinking about Brian's comment above.)